### ISSUE TITLE
FK Ids to show in Grid as Name Field instead of Integers.

### ISSUE THREAD

**Issue by stixoffire** [volkanceylan/Serenity#1553]
I have several fields in a table that have FK relations and are integers. The values show up just fine . Problem is I do not want the values, I would like the Corresponding Name field to show up. I have been looking at Northwind and samples (Order Table specifically as it has a ShipVia FK). I have created a formatter but now my field is blank (I can guess why it is blank - no data for the field.) 

So if I have table Project

Id
Description
EmployeeId_FK

My row will show in the grid:
1 DoStuff, 22

I would like it to show 
1, DoStuff, BillyBob T

My Formatter Code:
```
    @Serenity.Decorators.registerFormatter()
    export class EmployeeFormatter implements Slick.Formatter {
        format(ctx: Slick.FormatterContext) {
            let idList = ctx.value as number[];
            if (!idList || !idList.length)
                return "";
            let byId = EmployeeRow.getLookup().itemById;

          //  return Q.htmlEncode(byId[0].FirstName);
            return idList.map(x => {
                let g = byId[x];
                if (!g)
                    return x.toString();
                return Q.htmlEncode(g.FirstName);
            }).join(", ").toString();
        }
    }
```
 From my EmployeeRow.cs

```
//       I separated elements   in [] to individual lines for readability.
        [DisplayName("Employee"), ForeignKey("[dbo].[Employees]", "Id")]
        [LeftJoin("jEmployee"), TextualField("EmployeeFirstName"), Width(100)]   
        [LookupEditor(typeof(EmployeesRow))]
        public Int32? EmployeeId
        {
            get { return Fields.EmployeeId[this]; }
            set { Fields.EmployeeId[this] = value; }
        }

        [DisplayName("Employee Name"), Expression("jEmployee.[FirstName]")]
        [Width(100), QuickSearch]
        public String EmployeeFirstName
        {
            get { return Fields.FirstName[this]; }
            set { Fields.FirstName[this] = value; }
        }


```
I have since tried to accomplish this by using the following function:

Modified the  EmployeeFormatter: I now can see the Name field but clicking to bring up the Editor gives me an error - which I suspect is that I am now passing the name instead of the Id ,

```
    @Serenity.Decorators.registerFormatter()
    export class EmployeeFormatter implements Slick.Formatter {
        format(ctx: Slick.FormatterContext) {
            let EntId = ctx.value as number;
            if (EntId == null)
                return "";
            let ItembyId = EmployeeRow.getLookup().itemById;

            return Q.htmlEncode(ItembyId[EntId].FirstName.toString());

        }
    }
```

After Modification I receive the following :

Unhandled exception at line 619, column 21 in 
my local host /Scripts/Site/Jabber.Web.js?v=haO75OIIjWjE2YkYp7QL6A2

0x800a01b6 - JavaScript runtime error: Object doesn't support property or method 'map'

GridEditorBase

```
        public set value(value: TEntity[]) {
            var p = this.getIdProperty();
            this.view.setItems((value || []).map(x => {
                var y = Q.deepClone(x);
                if ((y as any)[p] == null)
                    (y as any)[p] = "`" + this.nextId++;
                return y;
            }), true);
        }
```

**Comment by volkanceylan** [comment]
You don't need such. Just do proper joins. See mapping guide.

**Comment by stixoffire** [comment]
@volkanceylan -   I let sergen generate the code for me originally and it simply showed the Id's in the form so I thought I had to change something - the documentation implied that when I selected the relationship mapping in sergen that this would generate the proper code for the foreign keys for me - which it looked like it did. 
 I will look at the mapping guide again.
 (I used 2.6.3 vsix 11/26/2016 ) to create the project in case that is useful).  I removed the formatter from my code , my code looks  like what is in the mapping guide [pgs 354-362, specifically pg's 358-362]. However I will look it over again to see what I am missing - should I be hiding these fk Id fields and making the expression fields some how Visible - I do not see this on the NorthWind Products and yet it works.

**Comment by volkanceylan** [comment]
Read the guide properly.

**Comment by stixoffire** [comment]
@volkenceylan 
After reading and working for the last couple days and also last night, I still see the same result - the FK Ids in the main grid are shown - the editor has the dropdowns like it should. and unless I add a formatter to the ProjectRow EmployeeId and CompanyId - I always see simply the Id - I can declare the textualfield or not declare it and I get the same result. I attached 2 images one with only the ids showing , the other how it should appear with the formatter , I do not have issue with using a formatter but if I do not need it - I can learn a better simpler way.

I have created a sample application and attached is my fluent migrations files DefaultDB really only the last one is needed to be added for DefaultDb) and I have included the built files for the testing project.

I also specifically duplicated how the Product Files from northwind were done (at least the cs files) since it works like I wanted mine to work. I replaced the supplier and category fields with my employee and company fields. I still see the same result unless I use a formatter.  Since addign the files here I noticed the Expression textual field had the same displayname property so I changed that to match and it still does not work with out the formatter. 

My joins are proper - I put a break point before save and I checked them - they show the left and the right hand assignments.  
Northwind Products page looks like on one ts file that there is a formatter selectFormatter  - so I commented out all that extra code and the product page still works right - yet mine does not. 
I would like to know what I am doing wrong - even though using a formatter works. (I have been bashing on this for 3 days plus)
 
![fk_idshowingtextualfield_using formatter](https://cloud.githubusercontent.com/assets/9748682/20862696/727a91f4-b977-11e6-9b61-cc4057fae18f.png)
![fk_idsnotshowingtextualfield](https://cloud.githubusercontent.com/assets/9748682/20862697/727dc7a2-b977-11e6-9344-99985b6a5119.png)
[Testing.zip](https://github.com/volkanceylan/Serenity/files/629170/Testing.zip)
[DefaultDB.zip](https://github.com/volkanceylan/Serenity/files/629169/DefaultDB.zip)

**Comment by volkanceylan** [comment]
Are you reading the tutorials?? Don't start work without reading tutorials. You change Columns.cs to property names of your text columns. Even at readme of GitHub repository, it is shown in the animation?

**Comment by stixoffire** [comment]
@volkanceylan long apology - I am really sorry..
Yes I had been reading the movie tutorials , that is how I got the project going and added different enumerated fields to my project and those things were all working - I read the movie genre tutorial which was right after the movie kind tutorial - thought it to be most similar. Now that you said it - I looked again between product in Northwind and my project - I was so fixated on the declarative programming that I missed it completely even in the tutorials and I have to say I am really really sorry - I hate bothering people for things that are written or can be figured out.  I went back to read again and I specifically looked for the columns and there it was very obvious pg 85 - it was even notated that way! 
I am really sorry for bothering you, I need to beat myself with a stick why I did not see that. I am really sorry..
